drivers/charger/sbs_charger.c

```diff
@@ -0,0 +1,190 @@
+{
+	const struct sbs_charger_config *config = dev->config;
+	uint8_t buf[2];
+
+	sys_put_le16(val, buf);
+
+	return i2c_burst_write_dt(&config->i2c, reg_addr, buf, sizeof(buf));
+}
+
+static int sbs_cmd_reg_update(const struct device *dev, uint8_t reg_addr, uint16_t mask,
+			      uint16_t val)
+{
+	uint16_t old_val, new_val;
+	int ret;
+
+	ret = sbs_cmd_reg_read(dev, SBS_CHARGER_REG_STATUS, &old_val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	new_val = (old_val & ~mask) | (val & mask);
+	if (new_val == old_val) {
+		return 0;
+	}
+
+	return sbs_cmd_reg_write(dev, reg_addr, new_val);
+}
+
+static int sbs_charger_get_prop(const struct device *dev, struct charger_get_property *prop)
+{
+	uint16_t reg_val;
+	int ret;
+
+	switch (prop->property_type) {
+	case CHARGER_PROP_ONLINE:
+		ret = sbs_cmd_reg_read(dev, SBS_CHARGER_REG_STATUS, &reg_val);
+		if (reg_val & SBS_CHARGER_STATUS_AC_PRESENT) {
+			prop->value.online = CHARGER_ONLINE_FIXED;
+		} else {
+			prop->value.online = CHARGER_ONLINE_OFFLINE;
+		}
+		break;
+	case CHARGER_PROP_PRESENT:
+		ret = sbs_cmd_reg_read(dev, SBS_CHARGER_REG_STATUS, &reg_val);
+		if (reg_val & SBS_CHARGER_STATUS_BATTERY_PRESENT) {
+			prop->value.present = true;
+		} else {
+			prop->value.present = false;
+		}
+		break;
+	case CHARGER_PROP_STATUS:
+		ret = sbs_cmd_reg_read(dev, SBS_CHARGER_REG_STATUS, &reg_val);
+		if (!(reg_val & SBS_CHARGER_STATUS_BATTERY_PRESENT)) {
+			prop->value.status = CHARGER_STATUS_NOT_CHARGING;
+		} else if (reg_val & SBS_CHARGER_STATUS_AC_PRESENT &&
+			   !(reg_val & SBS_CHARGER_STATUS_CHARGE_INHIBITED)) {
+			prop->value.status = CHARGER_STATUS_CHARGING;
+		} else {
+			prop->value.status = CHARGER_STATUS_DISCHARGING;
+		}
+		break;
+	default:
+		ret = -ENOTSUP;
+	}
+
+	prop->ret = ret;
+
+	return ret;
+}
+
+static int sbs_charger_set_prop(const struct device *dev, struct charger_set_property *prop)
+{
+	uint16_t reg_val = 0;
+	int ret;
+
+	switch (prop->property_type) {
+	case CHARGER_PROP_STATUS:
+		if (prop->value.status != CHARGER_STATUS_CHARGING) {
+			reg_val = SBS_CHARGER_MODE_INHIBIT_CHARGE;
+		}
+		ret = sbs_cmd_reg_update(dev, SBS_CHARGER_REG_CHARGER_MODE,
+					 SBS_CHARGER_MODE_INHIBIT_CHARGE, reg_val);
+		break;
+	default:
+		ret = -ENOTSUP;
+	}
+
+	prop->ret = ret;
+
+	return ret;
+}
+
+static int sbs_charger_get_props(const struct device *dev, struct charger_get_property *props,
+				 size_t len)
+{
+	int err_count = 0;
+
+	for (int i = 0; i < len; i++) {
+		int ret = sbs_charger_get_prop(dev, props + i);
+
+		err_count += ret ? 1 : 0;
+	}
+
+	err_count = (err_count == len) ? -1 : err_count;
+
+	return err_count;
+}
+
+static int sbs_charger_set_props(const struct device *dev, struct charger_set_property *props,
+				 size_t len)
+{
+	int err_count = 0;
+
+	for (int i = 0; i < len; i++) {
+		int ret = sbs_charger_set_prop(dev, props + i);
+
+		err_count += ret ? 1 : 0;
+	}
+
+	err_count = (err_count == len) ? -1 : err_count;
+
+	return err_count;
+}
```

Comment:
As discussed early in this review, this is precisely what each driver will need to replicate with the current API design (**BAD**). This was _fixed_, and we're back at the same place.

Reply:
Yeah this should go

Reply:
Sure. We can delay a multi-property API until it's been demonstrated as a necessity. The multi versus single property approach is not a cornerstone IMO.

Let's just go back to a single property here.

We can totally just wait for a multi property get/set to be requested if someone wants/needs it.

The whole fewer syscalls point is true, but it's not something I personally care about. That can be someone else's future request.

Reply:
The point is: it makes no sense for that code to be in the driver, it should be in the subsystem. The driver API should only cover get/set for a single property, if you then want to expand the API for handling multiple properties that's fine but that should be at subsystem level, not be implemented by each individual driver.

I could understand if the device had some specific api that would make fetching multiple properties more efficiently but that's not the case, as it stands this code would just get copy pasted to each individual driver.

Reply:
ack, I'll drop it. Someone else can put it together in the future.

Reply:
> ack, I'll drop it. Someone else can put it together in the future.

Just to be clear - the comment (from both Gerard and I) is not about not having that functionality, it's about implementing it outside of drivers, nothing wrong if you want to implement it at the subsystem level and have it use the single prop API internally, that's the way to go.